src/v/kafka/server/quota_manager.cc

```diff
@@ -346,27 +348,73 @@ ss::future<throttle_delay> quota_manager::throttle_fetch_tp(
 // they have not received any updates in ten window's worth of time.
 void quota_manager::gc() {
     auto full_window = _default_num_windows() * _default_window_width();
+    auto expire_threshold = clock::now() - 10 * full_window;
     ssx::background
-      = ssx::spawn_with_gate_then(
-          _gate,
-          [full_window, _client_quotas = std::ref(_client_quotas)]() {
-              auto now = clock::now();
-              auto expire_age = full_window * 10;
-              auto new_map = client_quotas_map_t{*_client_quotas.get().local()};
-              auto erased = absl::erase_if(
-                new_map,
-                [now,
-                 expire_age](const std::pair<ss::sstring, client_quota>& q) {
-                    return (now - q.second.last_seen) > expire_age;
-                });
-              if (erased) {
-                  return _client_quotas.get().reset(std::move(new_map));
-              }
-              return ss::now();
-          })
-          .handle_exception([](const std::exception_ptr& e) {
-              vlog(klog.warn, "Error garbage collecting quotas - {}", e);
-          });
+      = ssx::spawn_with_gate_then(_gate, [this, expire_threshold]() {
+            return do_gc(expire_threshold);
+        }).handle_exception([](const std::exception_ptr& e) {
+            vlog(klog.warn, "Error garbage collecting quotas - {}", e);
+        });
+}
+
+ss::future<> quota_manager::do_gc(clock::time_point expire_threshold) {
+    vassert(
+      ss::this_shard_id() == _client_quotas.shard_id(),
+      "do_gc() should only be called on the owner shard");
+
+    using key_set = chunked_vector<ss::sstring>;
+
+    auto mapper =
+      [this, expire_threshold](ss::shard_id shard) -> ss::future<key_set> {
+        return ss::smp::submit_to(shard, [this, expire_threshold]() {
+            auto res = key_set{};
+            auto map_shared_ptr = _client_quotas.local().get();
+            for (const auto& kv : *map_shared_ptr) {
+                auto last_seen_tp = kv.second.last_seen_ms->local();
+                if (last_seen_tp < expire_threshold) {
+                    res.push_back(kv.first);
+                }
+            }
+            // Note: need to pre-sort the vector for the std::set_intersection
+            // in the reduce step
+            std::sort(res.begin(), res.end());
+            return res;
+        });
+    };
+
+    auto reducer = [](const key_set& acc, const key_set& next) -> key_set {
+        key_set result;
+        // Note: std::set_intersection assumes the inputs are sorted and
+        // guarantees that the output is sorted
+        std::set_intersection(
+          acc.begin(),
+          acc.end(),
+          next.begin(),
+          next.end(),
+          std::back_inserter(result));
+        return result;
```

Comment:
It might be worth benchmarking this, there are potentially lots of allocs.

It might be cheaper to take the parameters by value (I think they're moved in), and remove from `acc` the values in `next`.

Reply:
Any word on this? Assuming there are copies made at the `reducer` call site (in `mp_reduce0`), I don't see why we wouldn't compute the intersection in place.

Reply:
Yeah I haven't yet benchmarked gc specifically, but passing by value and doing the intersection in-place makes sense. Let me take this up as a follow up performance improvement PR though to unblock Ben's work on dynamic quota changes.

Reply:
sgtm